### PR TITLE
fix: slow message insert from scala

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

_Briefly describe the issue you have solved or implemented with this pull request. If the PR contains multiple issues, use a bullet list._

### Causes (Optional)

_Briefly describe the causes behind the issues. This could be helpful to understand the adopted solutions behind some nasty bugs or complex issues._

### Solutions

_Briefly describe the solutions you have implemented for the issues explained above._

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/3/) **failed**.

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/4/) **failed**.

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/6/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5606203)(5606203)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5606203)(5606203)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [8](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/8/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5610793)(5610793)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5610793)(5610793)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [10](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/10/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5611202)(5611202)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5611202)(5611202)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [11](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/11/) **failed**.

**Comment by AndroidBob** [comment]
Build [12](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/12/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5612299)(5612299)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5612299)(5612299)-dev-compat.apk)

File: app/src/main/kotlin/com/wire/android/migration/feature/MigrateClientsDataUseCase.kt [hunk 1]
```diff
@@ -49,6 +49,10 @@ class MigrateClientsDataUseCase @Inject constructor(
     private val scalaUserDBProvider: ScalaUserDatabaseProvider,
     private val userDataStoreProvider: UserDataStoreProvider
 ) {
+
+    suspend operator fun invoke(userId: UserId, isFederated: Boolean): Either<CoreFailure, Unit> =
+        invoke(listOf(userId), isFederated).values.first()
+
     @Suppress("LoopWithTooManyJumpStatements", "ComplexMethod")
     suspend operator fun invoke(userIds: List<UserId>, isFederated: Boolean): Map<UserId, Either<CoreFailure, Unit>> {
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/migration/feature/MigrateClientsDataUseCase.kt]: There are no usages of `invoke` with a list. We can simplify the code by removing it and the loop thingy.

File: app/src/main/kotlin/com/wire/android/migration/feature/MigrateConversationsUseCase.kt [hunk 2]
```diff
@@ -38,15 +39,19 @@ class MigrateConversationsUseCase @Inject constructor(
     private val scalaUserDatabase: ScalaUserDatabaseProvider,
     private val mapper: MigrationMapper
 ) {
+
+    suspend operator fun invoke(userId: UserId): Either<CoreFailure, List<ScalaConversationData>> =
+        invoke(listOf(userId)).map { it[userId] ?: listOf() }
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/migration/feature/MigrateConversationsUseCase.kt:44]: Same here

File: app/src/main/kotlin/com/wire/android/migration/feature/MigrateMessagesUseCase.kt [hunk 3]
```diff
@@ -39,18 +40,27 @@ class MigrateMessagesUseCase @Inject constructor(
...
 ) {
 
-    suspend operator fun invoke(userIdsAndConversationIds: Map<UserId, List<ScalaConversationData>>): Either<CoreFailure, Unit> =
+    suspend operator fun invoke(
+        userId: UserId, scalaConversations: List<ScalaConversationData>,
+        coroutineScope: CoroutineScope
+    ): Either<CoreFailure, Unit> =
+        invoke(mapOf(userId to scalaConversations), coroutineScope)
+
+    suspend operator fun invoke(
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/migration/feature/MigrateMessagesUseCase.kt]: And here

File: app/src/main/kotlin/com/wire/android/migration/feature/MigrateUsersUseCase.kt [hunk 4]
```diff
@@ -44,11 +45,14 @@ class MigrateUsersUseCase @Inject constructor(
...
+                    mapper.fromScalaUserToUser(scalaUser, selfScalaUser.id, selfScalaUser.domain, selfScalaUser.teamId, userId)
                 }
                 val sessionScope = coreLogic.getSessionScope(userId)
                 sessionScope.users.persistMigratedUsers(mappedUsers)
                 Either.Right(acc + userId)
             } else Either.Right(acc)
         }
+
+    suspend operator fun invoke(userId: UserId): Either<CoreFailure, UserId> =
+        invoke(listOf(userId)).map { userId }
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/migration/feature/MigrateUsersUseCase.kt]: Anotha one

File: app/src/main/kotlin/com/wire/android/migration/userDatabase/ScalaMessageDAO.kt [hunk 5]
```diff
@@ -77,21 +80,30 @@ data class ScalaMessageData(
...
+            val accumulator = mutableListOf<ScalaMessageData>()
+            scalaConversations.forEach { scalaConversation ->
+                accumulator += messagesFromConversation(scalaConversation)
+            }
+            accumulator
+        }
+
+    suspend fun messagesForConversation(scalaConversation: ScalaConversationData): List<ScalaMessageData> =
+        withContext(queryContext) {
+            messagesFromConversation(scalaConversation)
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/migration/userDatabase/ScalaMessageDAO.kt]:
```suggestion
```

Not being used

File: app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt [hunk 6]
```diff
@@ -81,50 +93,64 @@ class MigrationManager @Inject constructor(
...
+                    val userid = (it.value as Either.Right).value
+                    appLogger.d("$TAG - Step 2 - Migrating clients for $userid")
+                    migrateClientsData(userid, isFederated).onFailure { failure ->
+                        appLogger.e("$TAG - Step 2 - Migrating clients for $userid failed reason $failure")
+                    }
+                    appLogger.d("$TAG - Step 3 - Migrating users for $userid")
+                    migrateUsers(userid).flatMap {
+                        migrateConversations(it)
+                    }.flatMap {
+                        appLogger.d("$TAG - Step 4 - Migrating conversations for $userid")
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt]:
```suggestion
                        appLogger.d("$TAG - Step 4 - Migrating conversations for $userid")
                        migrateConversations(it)
                    }.flatMap {
                        appLogger.d("$TAG - Step 5 - Migrating messages for $userid")
```

**Review by vitorhugods — CHANGES_REQUESTED** [review]
A couple of suggestions to simplify the code a bit, but looks good in general :)

**Review by vitorhugods — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [15](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/15/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5613355)(5613355)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5613355)(5613355)-dev-compat.apk)

**Comment by AndroidBob** [comment]
Build [16](https://10.10.124.134/job/wire-android-reloaded/job/PR-1402/16/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(5613518)(5613518)-dev-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/compat/com.wire.android-vv0.0.1(5613518)(5613518)-dev-compat.apk)

**Pull request merged**